### PR TITLE
[Core] Enable CUDA graphs for DP + All2All kernels 

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Enable CUDA Graphs for DP + All2All kernels.

Fixes:
 1. The input buffers to the quant_method aren't captured properly when using CUDAGraphs + torch.compile. This PR introduces a staging area where the hidden_states and router_logits are copied into and it is this tensor that gets passed into quant_method.
 2. It is important that all DP ranks invoke the same number of dispatch and combine kernels. The kernels need to synchronize between DP ranks. When this requirement isn't respected, it manifests as a deadlock. To this effect, introduce a `get_dp_padding` method in `gpu_model_runner.py`.
 
 Tests:
 Verified correctness  using lm_eval locally on 4xH100.

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -830,6 +830,21 @@ def __init__(
 
         self.quant_method.create_weights(layer=self, **moe_quant_params)
 
+        # Chunked all2all staging tensor
+        self.batched_hidden_states: Optional[torch.Tensor] = None
+        self.batched_router_logits: Optional[torch.Tensor] = None
+        if self.moe_parallel_config.use_pplx_kernels:
+            act_dtype = torch.get_default_dtype()
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py:837]: Is this always correct?  Can we make this a part of the moe config (even if we get it from the same place).

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1105,6 +1105,25 @@ def sync_and_slice_intermediate_tensors(
...
+        num_tokens_across_dp = [0] * dp_size
+        num_tokens_across_dp[dp_rank] = num_tokens
+        num_tokens_tensor = torch.tensor(num_tokens_across_dp,
+                                         device="cpu",
+                                         dtype=torch.int32)
+        from vllm.distributed.parallel_state import get_dp_group
+        torch.distributed.all_reduce(num_tokens_tensor,
+                                     group=get_dp_group().cpu_group)
+        max_tokens_across_dp_cpu = torch.max(num_tokens_tensor).item()
+        return max_tokens_across_dp_cpu - num_tokens
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply vllm/v1/worker/gpu_model_runner.py:1125]: This is called before the forward_context is set.

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/layer.py:837]: I am not sure if it is always correct. @mgoin is there a better way to get the activation dtype here ? any pointers ? Thanks.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlrmchlsmth) [reply vllm/v1/worker/gpu_model_runner.py:1125]: Looks like each DP rank is padding out to the maximum number of tokens? That's definitely too much padding in the chunked prefill case. Might be OK for disagg P/D. 

BTW, could you factor this bit of code out so we don't duplicate it both here and in the forward_context?

**Review by bnellnm — APPROVED** [review]

**Review by tlrmchlsmth — APPROVED** [review]
